govcd/service_account.go

```diff
@@ -0,0 +1,371 @@
+	if err != nil {
+		return fmt.Errorf("error granting service account: %s", err)
+	}
+
+	tenantContext, err := sa.org.getTenantContext()
+	if err != nil {
+		return fmt.Errorf("error granting service account: %s", err)
+	}
+
+	err = client.OpenApiPostItem("37.0", urlRef, nil, userCode, nil, getTenantContextHeader(tenantContext))
+	if err != nil {
+		return fmt.Errorf("error granting service account: %s", err)
+	}
+
+	return nil
+}
+
+// GetInitialApiToken gets the initial API token for the Service Account and sets it in `Active` status
+func (sa *ServiceAccount) GetInitialApiToken() (*types.ApiTokenRefresh, error) {
+	if sa.authParams == nil {
+		return nil, fmt.Errorf("error: service account must be authorized and granted")
+	}
+	client := sa.org.client
+	uuid := extractUuid(sa.ServiceAccount.ID)
+	data := bytes.NewBufferString(
+		fmt.Sprintf("grant_type=%s&client_id=%s&device_code=%s",
+			"urn:ietf:params:oauth:grant-type:device_code",
+			uuid,
+			sa.authParams.DeviceCode,
+		))
+
+	token, err := client.GetAccessToken(sa.ServiceAccount.Org.Name, "CreateServiceAccount", data)
+	if err != nil {
+		return nil, err
+	}
+	return token, nil
+}
+
+// Refresh updates the Service Account object
+func (sa *ServiceAccount) Refresh() error {
+	if sa.ServiceAccount == nil || sa.org.client == nil || sa.ServiceAccount.ID == "" {
+		return fmt.Errorf("cannot refresh Edge Gateway without ID")
+	}
+
+	updatedServiceAccount, err := sa.org.GetServiceAccountById(sa.ServiceAccount.ID)
+	if err != nil {
+		return err
+	}
+	sa.ServiceAccount = updatedServiceAccount.ServiceAccount
+
+	return nil
+}
+
+// Revoke revokes the service account and its' API token and puts it back in 'Created' stage
+func (sa *ServiceAccount) Revoke() error {
+	client := sa.org.client
+
+	endpoint := types.OpenApiPathVersion1_0_0 + types.OpenApiEndpointServiceAccounts
+	apiVersion, err := client.getOpenApiHighestElevatedVersion(endpoint)
+	if err != nil {
+		return err
+	}
+
+	urlRef, err := client.OpenApiBuildEndpoint(endpoint, sa.ServiceAccount.ID, "/revoke")
+	if err != nil {
+		return err
+	}
+
+	err = client.OpenApiPostItem(apiVersion, urlRef, nil, nil, nil, nil)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Delete deletes a Service Account
+func (sa *ServiceAccount) Delete() error {
+	client := sa.org.client
+	if client.APIVCDMaxVersionIs("< 37.0") {
+		version, err := client.GetVcdFullVersion()
+		if err == nil {
+			return fmt.Errorf("minimum version for Service Accounts is 10.4.0 - Version detected: %s", version.Version)
+		}
+		// If we can't get the VCD version, we return API version info
+		return fmt.Errorf("minimum API version for Service Accounts is 37.0 - Version detected: %s", client.APIVersion)
+	}
+
+	endpoint := types.OpenApiPathVersion1_0_0 + types.OpenApiEndpointServiceAccounts
+	urlRef, err := client.OpenApiBuildEndpoint(endpoint, sa.ServiceAccount.ID)
+	if err != nil {
+		return err
+	}
+
+	err = client.OpenApiDeleteItem("37.0", urlRef, nil, nil)
```

Comment:
You could leverage `endpointMinApiVersions` structure and use `getOpenApiHighestElevatedVersion` - this will avoid having hardcoded API version in a separate place and also do the version check automatically that will make this part unneeded https://github.com/vmware/go-vcloud-director/pull/577/files#diff-3743a8e278e471adff2ca7110df5cc98b9600165b09face7e71dd9bfef78ea54R319-R327

Just like in the other place you have done - https://github.com/vmware/go-vcloud-director/pull/577/files#diff-3743a8e278e471adff2ca7110df5cc98b9600165b09face7e71dd9bfef78ea54R299

Reply:
Fixed.